orderer/common/follower/block_puller.go

```diff
@@ -24,73 +24,152 @@ import (
 type ChannelPuller interface {
 	PullBlock(seq uint64) *common.Block
 	HeightsByEndpoints() (map[string]uint64, error)
+	UpdateEndpoints(endpoints []cluster.EndpointCriteria)
 	Close()
 }
 
-// BlockPullerFromJoinBlock creates a new block puller from a join-block.
-// This block puller is used for on-boarding, when join-block.Number >= ledger.Height().
-// This block puller can fetch the genesis block and skip verifying it.
-func BlockPullerFromJoinBlock(
-	joinBlock *common.Block,
+//go:generate counterfeiter -o mocks/block_puller_factory.go -fake-name BlockPullerFactory . BlockPullerFactory
+
+type BlockPullerFactory interface {
+	BlockPuller(configBlock *common.Block) (ChannelPuller, error)
+	UpdateVerifierFromConfigBlock(configBlock *common.Block) error
+}
+
+// BlockPullerCreator creates a BlockPuller on demand.
+// It also maintains a link to a block signature verifier, and exposes a method to update it on incoming config blocks.
+// The BlockPuller generated by this factory always accesses the updated verifier, since it is generated
+// with a link to the factory's VerifyBlockSequence method.
+type BlockPullerCreator struct {
+	channelID               string
+	bccsp                   bccsp.BCCSP
+	blockSigVerifierFactory cluster.VerifierFactory // Creates a new block signature verifier
+	blockSigVerifier        cluster.BlockVerifier   // The current block signature verifier, from the latest channel config
+	clusterConfig           localconfig.Cluster
+	signer                  identity.SignerSerializer
+	der                     *pem.Block
+	stdDialer               *cluster.StandardDialer
+	clusterVerifyBlocks     ClusterVerifyBlocksFunc // Default: cluster.VerifyBlocks, or a mock for testing
+}
+
+// ClusterVerifyBlocksFunc is a function that matches the signature of cluster.VerifyBlocks, and allows mocks for testing.
+type ClusterVerifyBlocksFunc func(blockBuff []*common.Block, signatureVerifier cluster.BlockVerifier) error
+
+// NewBlockPullerFactory creates a new BlockPullerFactory, using the configuration details that do not change during
+// the life cycle of the orderer.
+func NewBlockPullerFactory(
 	channelID string,
+	logger *flogging.FabricLogger,
 	signer identity.SignerSerializer,
-	blockVerifier cluster.BlockVerifier,
 	baseDialer *cluster.PredicateDialer,
 	clusterConfig localconfig.Cluster,
 	bccsp bccsp.BCCSP,
-) (*cluster.BlockPuller, error) {
+	alternativeVerifyBlocksFunc ...ClusterVerifyBlocksFunc,
+) (BlockPullerFactory, error) {
 	if baseDialer == nil {
 		return nil, errors.New("base dialer is nil")
```

Comment:
I realize this is not changed here, but checking arguments for nil is also a bit odd.  Generally speaking nil is not a valid parameter unless explicitly documented to be so.  Looking at the stdlib you can find lots of places that take pointers and simply crash if they are invalid.

Reply:
done